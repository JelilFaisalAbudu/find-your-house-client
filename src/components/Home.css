.hero {
  background: url('../images/hero.jpg') no-repeat center center fixed;
  background-attachment: initial;
  -webkit-background-size: 100%;
  -moz-background-size: 100%;
  -o-background-size: 100%;
  background-size: 100%;
}

.hero-text {
  color: #fff;
}

.house-info {
  position: absolute;
  bottom: 0;
  margin-left: 1em;
}

.house .house-info,
.house .btn {
  transition: display 2s ease-in-out;
  display: none;
  z-index: 10;
}

.house:hover .house-info,
.house:hover .btn {
  transition: display 2s ease-in-out;
  display: unset;
}

.house img {
  border-radius: 0.5em;
  box-shadow: 0 0.46875rem 2.1875rem rgba(4, 9, 20, 0.03), 0 0.9375rem 1.40625rem rgba(4, 9, 20, 0.03), 0 0.25rem 0.53125rem rgba(4, 9, 20, 0.05), 0 0.125rem 0.1875rem rgba(4, 9, 20, 0.03);
}

.house:hover img {
  filter: brightness(0.6);
}
