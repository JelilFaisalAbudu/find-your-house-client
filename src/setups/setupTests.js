import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

Enzyme.configure({ adapter: new Adapter() });
