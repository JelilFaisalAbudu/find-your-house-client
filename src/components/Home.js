/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import history from '../helpers/history';
import { spyHeaderScroll } from '../helpers/dom';

import houseImage from '../images/scott-webb-1ddol8rgUH8-unsplash.jpg';
import userFavorites from '../redux/actions/favorite';
import getHouses from '../redux/actions/houses';
import './Home.css';

const Home = () => {
  const [loading, setLoading] = useState(false);
  const dispatch = useDispatch();
  const { user } = useSelector(state => state.auth);
  const { isLoggedIn } = useSelector(state => state.auth);
  const { favoriteHousesIds } = useSelector(state => state.favorites);
  const { houses } = useSelector(state => state.houses);
  const { networkErrorMessage } = useSelector(state => state.message);

  const getUserFavorites = userId => {
    dispatch(userFavorites.getUserFavorites(userId));
  };

  const handleAddFavorite = (userId, houseId) => {
    dispatch(userFavorites.addUserFavorite(userId, houseId));
  };

  const handleRemoveFavorite = (userId, houseId) => {
    dispatch(userFavorites.removeUserFavorite(userId, houseId));
    dispatch(userFavorites.doRemoveFavorite(houseId));
  };

  const changeRoute = (path = '/sigin') => history.replace(`${path}`);

  const displayFavoriteBtn = houseId => {
    if (isLoggedIn && favoriteHousesIds.includes(houseId)) {
      return (
        <button
          type="button"
          aria-label="Remove from favorite list"
          className="btn btn-unlike"
          data-toggle="tooltip"
          data-placement="top"
          title="remove from favorite list"
          onClick={() => handleRemoveFavorite(user.id, houseId)}
        >
          <span className="material-icons">
            favorite
          </span>
        </button>
      );
    }

    return (
      <button
        type="button"
        aria-label="Add to favorite list"
        className="btn btn-like"
        data-toggle="tooltip"
        data-placement="top"
        title="add to favorite list"
        onClick={() => {
          if (!isLoggedIn) {
            return changeRoute('/signin');
          }
          return handleAddFavorite(user.id, houseId);
        }}
      >
        <span className="material-icons">
          favorite
        </span>
      </button>
    );
  };

  useEffect(() => {
    setLoading(true);
    dispatch(getHouses());
    setLoading(false);
    spyHeaderScroll(400);
  }, [spyHeaderScroll]);

  useEffect(() => {
    if (user) {
      getUserFavorites(user.id);
    }
  }, []);

  let content;

  if (networkErrorMessage) {
    content = (
      <header className="jumbotron">
        <h4>{networkErrorMessage}</h4>
      </header>
    );
  } else {
    content = (

      <>
        <div className="hero position-relative overflow-hidden p-3 p-md-5 text-center bg-light">
          <div className="hero-text col-md-5 p-lg-5 mx-auto my-5">
            <h1 className="display-4 fw-normal">Punny headline</h1>
            <p className="lead fw-normal">A landing page to boot. Jumpstart your marketing efforts with this example based on  marketing pages.</p>
            <a className="btn btn-outline-secondary" to="#">Coming soon</a>
          </div>
        </div>

        <div className="container-fluid">
          <div className="houses-container py-2 py-md-4 row g-3">
            {houses.length
              && (houses.map(house => (
                <div key={house.id} className="house col-12 col-lg-6 col-xlg-4 text-white position-relative">
                  {displayFavoriteBtn(house.id)}
                  <div className="house-info">
                    <h2 className="display-6">{house.name}</h2>
                    <h3 className="display-7">{house.category}</h3>
                    <p className="lead">And an even wittier subheading.</p>
                  </div>
                  <img
                    className="img-fluid"
                    src={houseImage}
                    alt={house.name}
                  />
                </div>

              )))}
          </div>
        </div>

      </>
    );
  }

  return loading ? <div>Loading...</div> : content;
};

export default Home;
