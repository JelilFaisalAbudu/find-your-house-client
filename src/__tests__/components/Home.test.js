import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import '../../setups/setupTests';
import Home from '../../components/Home';
import store from '../../redux/store';

const setup = () => mount(
  <Provider store={store}>
    <Home />
  </Provider>,
);

describe('Home', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  test('renders without crashing', () => {
    // console.log(toJson(wrapper));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('', () => {
    const houses = [
      {
        id: 1,
        name: 'Golden Lu',
        category: 'Glass House',
        description: 'Lorem Ipsum',
        imageUrl: 'http://example.com/golden-lu.jpg',
      },
      {
        id: 2,
        name: 'Golden Lu',
        category: 'Glass House',
        description: 'Lorem Ipsum',
        imageUrl: 'http://example.com/golden-lu.jpg',
      },
    ];
    const action = houses => ({
      type: 'GET_HOUSES',
      payload: { houses },
    });

    beforeAll(() => {
      store.dispatch(action(houses));
    });

    test('renders the Home form correctly', async () => {
      const renderedHouses = await wrapper.find('.houses-container').children();
      expect(renderedHouses.length).toEqual(houses.length);
    });
  });
});
