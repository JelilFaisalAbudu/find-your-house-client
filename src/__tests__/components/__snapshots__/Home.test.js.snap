// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders without crashing 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Home>
    <div
      className="hero position-relative overflow-hidden p-3 p-md-5 text-center bg-light"
    >
      <div
        className="hero-text col-md-5 p-lg-5 mx-auto my-5"
      >
        <h1
          className="display-4 fw-normal"
        >
          Punny headline
        </h1>
        <p
          className="lead fw-normal"
        >
          A landing page to boot. Jumpstart your marketing efforts with this example based on  marketing pages.
        </p>
        <a
          className="btn btn-outline-secondary"
          to="#"
        >
          Coming soon
        </a>
      </div>
    </div>
    <div
      className="container-fluid"
    >
      <div
        className="houses-container py-2 py-md-4 row g-3"
      >
        0
      </div>
    </div>
  </Home>
</Provider>
`;
